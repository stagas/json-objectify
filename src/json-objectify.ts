import { asyncSerialReduce } from 'everyday-utils'

interface Replacer {
  (this: any, key: string, value: any): any
}

interface Reviver {
  (this: any, key: string, value: any): any
}

export const createContext = (asIsCtors: any[] = []) => {
  const asIs = new Set(asIsCtors)

  const objectify = (value: any, replacer: Replacer, top = true): any => {
    if (value === null) return value

    if (typeof value === 'object' && !(asIs.has(value.constructor))) {
      if (top) {
        value = replacer.call({ '': value }, '', value)
      }

      value = Object.entries(value).reduce((p, [k, v]) => {
        p[k] = v === null
          ? v
          : objectify(replacer.call(value, k, v), replacer, false)
        return p
      }, Array.isArray(value) ? [] : {} as any)

      return value
    }

    return value
  }

  const objectifyAsync = async (
    value: any,
    replacer: Replacer,
    top = true,
  ): Promise<any> => {
    if (value === null) return value

    if (typeof value === 'object' && !(asIs.has(value.constructor))) {
      if (top) {
        value = replacer.call({ '': value }, '', value)
      }

      value = await asyncSerialReduce(Object.entries(value), async (p, [k, v]) => {
        p[k] = v === null
          ? v
          : await new Promise(resolve => {
            setTimeout(async () => {
              resolve(
                await objectifyAsync(replacer.call(value, k, v), replacer, false)
              )
            })
          })
        return p
      }, Array.isArray(value) ? [] : {} as any)

      return value
    }

    return value
  }

  const deobjectify = (value: any, reviver: Reviver, top = true): any => {
    if (value === null) return value

    if (typeof value === 'object' && !(asIs.has(value.constructor))) {
      value = Object.entries(value).reduce((p, [k, v]) => {
        p[k] = v === null
          ? v
          : reviver.call(p, k, deobjectify(v, reviver, false))
        return p
      }, Array.isArray(value) ? [] : {} as any)

      if (top) {
        value = reviver.call({ '': value }, '', value)
      }

      return value
    }

    return value
  }

  return { objectify, objectifyAsync, deobjectify }
}

export const { objectify, objectifyAsync, deobjectify } = createContext()
