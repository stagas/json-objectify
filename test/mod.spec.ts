import { createContext as createSerializeContext, replacer, reviver, thru } from 'serialize-whatever'
import { createContext as createObjectifyContext, deobjectify, objectify, objectifyAsync } from '../src/json-objectify'

describe('objectify(obj)', () => {
  it('works', () => {
    const obj = { foo: 'bar' }
    const res = objectify(obj, replacer(obj))
    expect(res).toMatchSnapshot()
  })

  it('SharedArrayBuffer', () => {
    const { replacer, reviver } = createSerializeContext([
      [Float64Array, [thru, thru]],
    ])
    const { objectify, deobjectify } = createObjectifyContext([Float64Array])
    const obj = {
      foo: new Float64Array(
        new SharedArrayBuffer(Float64Array.BYTES_PER_ELEMENT)
      ),
    }
    const str = objectify(obj, replacer(obj))
    expect(str).toMatchSnapshot()
    const res = deobjectify<any>(str, reviver([]))
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res.foo === obj.foo).toBe(true)
  })

  it('ref SharedArrayBuffer', () => {
    const { replacer, reviver } = createSerializeContext([
      [Float64Array, [thru, thru]],
    ])
    const { objectify, deobjectify } = createObjectifyContext([Float64Array])
    const obj = {
      foo: new Float64Array(
        new SharedArrayBuffer(Float64Array.BYTES_PER_ELEMENT)
      ),
    } as any
    obj.fooRef = obj.foo
    const str = objectify(obj, replacer(obj))
    expect(str).toMatchSnapshot()
    const res = deobjectify<any>(str, reviver([]))
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res.foo === obj.foo).toBe(true)
    expect(res.foo === obj.fooRef).toBe(true)
    expect(res.fooRef === obj.foo).toBe(true)
    expect(res.fooRef === obj.fooRef).toBe(true)
  })

  it('arrays', () => {
    const obj = [1, 'foo', null]
    const res = objectify(obj, replacer(obj))
    expect(res).toMatchSnapshot()
  })

  it('complex', () => {
    const obj = [{
      a: 123,
      b: new Date(123),
      c: new Map([[1, 2], [3, 4]]),
      d: new Set([1, 2, 3, 4]),
    }]

    const res = objectify(obj, replacer(obj))
    expect(res).toMatchSnapshot()
  })

  it('cross refs', () => {
    const foo = { hello: 'world', bar: undefined } as any
    const bar = { one: 'two', b: foo }
    foo.bar = bar
    const obj = {
      a: foo,
      b: bar,
    } as any
    obj.c = obj.a
    obj.d = { a: obj.a, b: obj.b }

    const str = objectify(obj, replacer(obj))
    expect(str).toMatchSnapshot()
    const res = deobjectify(str, reviver([]))
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('refs in maps and sets', () => {
    const foo = { hello: 'world' }
    const arr = [1, 2, 3]
    const arrRef = [1, 2, foo]
    const obj = {
      a: foo,
      b: new Date(123),
      c: new Map([[1, foo], [2, arr]] as any),
      d: new Set([1, 2, foo, 4]),
      e: arr,
      f: arrRef,
    }
    const str = objectify(obj, replacer(obj))
    expect(str).toMatchSnapshot()
    const res = deobjectify(str, reviver([])) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res.c.get(1)).toBe(res.a)
    expect(res.c.get(2)).toBe(res.e)
    expect(res.f[2]).toBe(res.a)
  })

  it('refs in maps keys', () => {
    const foo = { hello: 'world' }
    const bar = { one: 'two' }
    const obj = {
      a: new Map([[foo, bar]]),
    }
    const str = objectify(obj, replacer(obj))
    expect(str).toMatchSnapshot()
    const res = deobjectify(str, reviver([])) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('preserves references', () => {
    class Foo {
      a = 1
      b = 2
      constructor(state: Partial<Foo> = {}) {
        Object.assign(this, state)
      }
      toJSON() {
        return {
          a: this.a,
        }
      }
    }
    const foo = new Foo({ a: 2 })
    const obj = {
      a: 123,
      b: foo,
      c: new Date(123),
      d: foo,
    }
    expect(obj).toMatchSnapshot()

    const str = objectify(obj, replacer(obj))

    expect(str).toMatchSnapshot()
    const res = deobjectify(str, reviver([Foo])) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res.d).toBe(res.b)
    expect(res.b).toBeInstanceOf(Foo)
  })

  it('toJSON top', () => {
    class Foo {
      a = 1
      b = 2
      constructor(state: Partial<Foo> = {}) {
        Object.assign(this, state)
      }
      toJSON() {
        return {
          a: this.a,
        }
      }
    }
    const obj = new Foo({ a: 2 })
    const str = objectify(obj, replacer(obj))

    expect(str).toMatchSnapshot()
    const res = deobjectify(str, reviver([Foo])) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res).toBeInstanceOf(Foo)
  })

  it('async complex', async () => {
    const foo = { hello: 'world' }
    const arr = [1, 2, 3]
    const arrRef = [1, 2, foo]
    const obj = {
      a: foo,
      b: new Date(123),
      c: new Map([[1, foo], [2, arr]] as any),
      d: new Set([1, 2, foo, 4]),
      e: arr,
      f: arrRef,
    }
    const str = await objectifyAsync(obj, replacer(obj))
    expect(str).toMatchSnapshot()
    const res = deobjectify(str, reviver([])) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res.c.get(1)).toBe(res.a)
    expect(res.c.get(2)).toBe(res.e)
    expect(res.f[2]).toBe(res.a)
  })
})
