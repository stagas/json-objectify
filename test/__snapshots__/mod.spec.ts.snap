
exports[`objectify(obj) works 1`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "foo": "bar",
  },
}
`;

exports[`objectify(obj) SharedArrayBuffer 1`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "foo": Object {
      "$r": 2,
      "$t": "Float64Array",
      "$v": Float64Array [
        0,
      ],
    },
  },
}
`;

exports[`objectify(obj) SharedArrayBuffer 2`] = `
Object {
  "foo": Float64Array [
    0,
  ],
}
`;

exports[`objectify(obj) ref SharedArrayBuffer 1`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "foo": Object {
      "$r": 2,
      "$t": "Float64Array",
      "$v": Float64Array [
        0,
      ],
    },
    "fooRef": Object {
      "$p": 2,
    },
  },
}
`;

exports[`objectify(obj) ref SharedArrayBuffer 2`] = `
Object {
  "foo": Float64Array [
    0,
  ],
  "fooRef": Float64Array [
    0,
  ],
}
`;

exports[`objectify(obj) arrays 1`] = `
Object {
  "$r": 1,
  "$t": "Array",
  "$v": Array [
    1,
    "foo",
    null,
  ],
}
`;

exports[`objectify(obj) complex 1`] = `
Object {
  "$r": 1,
  "$t": "Array",
  "$v": Array [
    Object {
      "$r": 2,
      "$t": "Object",
      "$v": Object {
        "a": 123,
        "b": Object {
          "$r": 3,
          "$t": "Date",
          "$v": "1970-01-01T00:00:00.123Z",
        },
        "c": Object {
          "$r": 4,
          "$t": "Map",
          "$v": Array [
            Object {
              "$r": 5,
              "$t": "Array",
              "$v": Array [
                1,
                2,
              ],
            },
            Object {
              "$r": 6,
              "$t": "Array",
              "$v": Array [
                3,
                4,
              ],
            },
          ],
        },
        "d": Object {
          "$r": 7,
          "$t": "Set",
          "$v": Array [
            1,
            2,
            3,
            4,
          ],
        },
      },
    },
  ],
}
`;

exports[`objectify(obj) cross refs 1`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "a": Object {
      "$r": 2,
      "$t": "Object",
      "$v": Object {
        "bar": Object {
          "$r": 3,
          "$t": "Object",
          "$v": Object {
            "b": Object {
              "$p": 2,
            },
            "one": "two",
          },
        },
        "hello": "world",
      },
    },
    "b": Object {
      "$p": 3,
    },
    "c": Object {
      "$p": 2,
    },
    "d": Object {
      "$r": 4,
      "$t": "Object",
      "$v": Object {
        "a": Object {
          "$p": 2,
        },
        "b": Object {
          "$p": 3,
        },
      },
    },
  },
}
`;

exports[`objectify(obj) cross refs 2`] = `
Object {
  "a": Object {
    "bar": Object {
      "b": [Circular],
      "one": "two",
    },
    "hello": "world",
  },
  "b": Object {
    "b": [Circular],
    "one": "two",
  },
  "c": Object {
    "bar": [Circular],
    "hello": "world",
  },
  "d": Object {
    "a": [Circular],
    "b": [Circular],
  },
}
`;

exports[`objectify(obj) refs in maps and sets 1`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "a": Object {
      "$r": 2,
      "$t": "Object",
      "$v": Object {
        "hello": "world",
      },
    },
    "b": Object {
      "$r": 3,
      "$t": "Date",
      "$v": "1970-01-01T00:00:00.123Z",
    },
    "c": Object {
      "$r": 4,
      "$t": "Map",
      "$v": Array [
        Object {
          "$r": 5,
          "$t": "Array",
          "$v": Array [
            1,
            Object {
              "$p": 2,
            },
          ],
        },
        Object {
          "$r": 6,
          "$t": "Array",
          "$v": Array [
            2,
            Object {
              "$r": 7,
              "$t": "Array",
              "$v": Array [
                1,
                2,
                3,
              ],
            },
          ],
        },
      ],
    },
    "d": Object {
      "$r": 8,
      "$t": "Set",
      "$v": Array [
        1,
        2,
        Object {
          "$p": 2,
        },
        4,
      ],
    },
    "e": Object {
      "$p": 7,
    },
    "f": Object {
      "$r": 9,
      "$t": "Array",
      "$v": Array [
        1,
        2,
        Object {
          "$p": 2,
        },
      ],
    },
  },
}
`;

exports[`objectify(obj) refs in maps and sets 2`] = `
Object {
  "a": Object {
    "hello": "world",
  },
  "b": 1970-01-01T00:00:00.123Z,
  "c": Map {
    1 => [Circular],
    2 => Array [
      1,
      2,
      3,
    ],
  },
  "d": Set {
    1,
    2,
    [Circular],
    4,
  },
  "e": Array [
    1,
    2,
    3,
  ],
  "f": Array [
    1,
    2,
    [Circular],
  ],
}
`;

exports[`objectify(obj) refs in maps keys 1`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "a": Object {
      "$r": 2,
      "$t": "Map",
      "$v": Array [
        Object {
          "$r": 3,
          "$t": "Array",
          "$v": Array [
            Object {
              "$r": 4,
              "$t": "Object",
              "$v": Object {
                "hello": "world",
              },
            },
            Object {
              "$r": 5,
              "$t": "Object",
              "$v": Object {
                "one": "two",
              },
            },
          ],
        },
      ],
    },
  },
}
`;

exports[`objectify(obj) refs in maps keys 2`] = `
Object {
  "a": Map {
    Object {
      "hello": "world",
    } => Object {
      "one": "two",
    },
  },
}
`;

exports[`objectify(obj) preserves references 1`] = `
Object {
  "a": 123,
  "b": Object {
    "a": 2,
  },
  "c": 1970-01-01T00:00:00.123Z,
  "d": Object {
    "a": 2,
  },
}
`;

exports[`objectify(obj) preserves references 2`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "a": 123,
    "b": Object {
      "$r": 2,
      "$t": "Foo",
      "$v": Object {
        "a": 2,
      },
    },
    "c": Object {
      "$r": 3,
      "$t": "Date",
      "$v": "1970-01-01T00:00:00.123Z",
    },
    "d": Object {
      "$p": 2,
    },
  },
}
`;

exports[`objectify(obj) preserves references 3`] = `
Object {
  "a": 123,
  "b": Object {
    "a": 2,
  },
  "c": 1970-01-01T00:00:00.123Z,
  "d": Object {
    "a": 2,
  },
}
`;

exports[`objectify(obj) toJSON top 1`] = `
Object {
  "$r": 1,
  "$t": "Foo",
  "$v": Object {
    "a": 2,
  },
}
`;

exports[`objectify(obj) toJSON top 2`] = `
Object {
  "a": 2,
}
`;

exports[`objectify(obj) async complex 1`] = `
Object {
  "$r": 1,
  "$t": "Object",
  "$v": Object {
    "a": Object {
      "$r": 2,
      "$t": "Object",
      "$v": Object {
        "hello": "world",
      },
    },
    "b": Object {
      "$r": 3,
      "$t": "Date",
      "$v": "1970-01-01T00:00:00.123Z",
    },
    "c": Object {
      "$r": 4,
      "$t": "Map",
      "$v": Array [
        Object {
          "$r": 5,
          "$t": "Array",
          "$v": Array [
            1,
            Object {
              "$p": 2,
            },
          ],
        },
        Object {
          "$r": 6,
          "$t": "Array",
          "$v": Array [
            2,
            Object {
              "$r": 7,
              "$t": "Array",
              "$v": Array [
                1,
                2,
                3,
              ],
            },
          ],
        },
      ],
    },
    "d": Object {
      "$r": 8,
      "$t": "Set",
      "$v": Array [
        1,
        2,
        Object {
          "$p": 2,
        },
        4,
      ],
    },
    "e": Object {
      "$p": 7,
    },
    "f": Object {
      "$r": 9,
      "$t": "Array",
      "$v": Array [
        1,
        2,
        Object {
          "$p": 2,
        },
      ],
    },
  },
}
`;

exports[`objectify(obj) async complex 2`] = `
Object {
  "a": Object {
    "hello": "world",
  },
  "b": 1970-01-01T00:00:00.123Z,
  "c": Map {
    1 => [Circular],
    2 => Array [
      1,
      2,
      3,
    ],
  },
  "d": Set {
    1,
    2,
    [Circular],
    4,
  },
  "e": Array [
    1,
    2,
    3,
  ],
  "f": Array [
    1,
    2,
    [Circular],
  ],
}
`;
